src/engine/Engine.cpp

```diff
@@ -3106,4 +3202,391 @@ InputQuery Engine::buildQueryFromCurrentState() const {
+        {
+            infVar = curBasicVar;
+            break;
+        }
+    }
+
+    delete costRow;
+    return infVar;
+}
+
+bool Engine::explainAndCheckContradiction( unsigned var, bool isUpper, const TableauRow *row )
+{
+    ASSERT( _produceUNSATProofs );
+
+    Vector<double> backup( 0 );
+    _boundManager.getExplanation( var, isUpper, backup );
+
+    _boundManager.updateBoundExplanation( *row, isUpper, var );
+
+    // Ensure the proof is correct
+    if ( certifyInfeasibility( var ) )
+        return true;
+
+    // If not, restores previous certificate if the proof is wrong
+    _boundManager.setExplanation( backup, var, isUpper );
+
+    return false;
+}
+
+bool Engine::explainAndCheckContradiction( unsigned var, bool isUpper, const SparseUnsortedList *row )
+{
+    ASSERT( _produceUNSATProofs );
+
+    Vector<double> backup( 0 );
+    _boundManager.getExplanation( var, isUpper, backup );
+
+    _boundManager.updateBoundExplanationSparse( *row, isUpper, var );
+
+    // Ensure the proof is correct
+    if ( certifyInfeasibility( var ) )
+        return true;
+
+    // If not, restores previous certificate if the proof is wrong
+    _boundManager.setExplanation( backup, var, isUpper );
+
+    return false;
+}
+
+ UnsatCertificateNode* Engine::getUNSATCertificateCurrentPointer() const
+{
+    return _UNSATCertificateCurrentPointer->get();
+}
+
+void Engine::setUNSATCertificateCurrentPointer( UnsatCertificateNode* node )
+{
+    _UNSATCertificateCurrentPointer->set( node );
+}
+
+const UnsatCertificateNode* Engine::getUNSATCertificateRoot() const
+{
+    return _UNSATCertificate;
+}
+
+bool Engine::certifyUNSATCertificate()
+{
+    ASSERT( _produceUNSATProofs && _UNSATCertificate && !_smtCore.getStackDepth() );
+
+    for ( auto &constraint : _plConstraints )
+    {
+        if ( constraint->getType() != RELU )
+        {
+            printf("Certification Error! Marabou doesn't support certification for none ReLU constraints.\n");
+            return false;
+        }
+    }
+
+    struct timespec certificationStart = TimeUtils::sampleMicro();
+    _precisionRestorer.restoreInitialEngineState( *this );
+
+    Vector<double> groundUpperBounds( _tableau->getN(), 0 );
+    Vector<double> groundLowerBounds( _tableau->getN(), 0 );
+
+    for ( unsigned i = 0; i < _tableau->getN(); ++i )
+    {
+        groundUpperBounds[i] = _groundBoundManager.getUpperBound( i );
+        groundLowerBounds[i] = _groundBoundManager.getLowerBound( i );
+    }
+
+    Checker unsatCertificateChecker( _UNSATCertificate, _tableau->getM(),  _tableau->getSparseA(),
+                                     groundUpperBounds, groundLowerBounds, _plConstraints );
+    bool certificationSucceeded = unsatCertificateChecker.check();
+
+    _statistics.setLongAttribute( Statistics::TOTAL_CERTIFICATION_TIME, TimeUtils::timePassed( certificationStart, TimeUtils::sampleMicro() ) );
+    printf( "Certification time: " );
```

Comment:
Should these really be printfs?

Reply:
I think It'll be nice to present it to the user (it doesn't add much, and parallel stats are presented).